File: design-pattern-python/design_pattern_python/observer/observer.py
import time
from abc import ABCMeta, abstractmethod

from number_generator import NumberGenerator, RandomNumberGenerator


class Observer(metaclass=ABCMeta):
    @abstractmethod
    def update(self, generator: NumberGenerator) -> None:
        pass


class DigitObserver(Observer):
    def update(self, generator: NumberGenerator) -> None:
        print(f"DigitObserver:{generator.get_number()}")
        time.sleep(0.1)


class GraphObserver(Observer):
    def update(self, generator: NumberGenerator) -> None:
        print("GraphObserver", end="")
        count = generator.get_number()
        for _ in range(count):
            print("*", end="")
        print("")
        time.sleep(0.1)


if __name__ == "__main__":
    generator = RandomNumberGenerator()
    observer1 = DigitObserver()
    observer2 = GraphObserver()
    generator.add_observer(observer1)
    generator.add_observer(observer2)
    generator.execute()

File: design-pattern-go/observer/observer.go
package observer

import (
	"fmt"
	"time"
)

type Observer interface {
	Update(generator INumberGenerator)
}

type DigitObserver struct {
	concreteObserver Observer
}

func NewDigitObserver() *DigitObserver {
	digitObserver := &DigitObserver{}
	digitObserver.concreteObserver = digitObserver

	return digitObserver
}

func (d *DigitObserver) Update(generator INumberGenerator) {
	fmt.Printf("DigitObserver:%d\n", generator.GetNumber())
	time.Sleep(time.Microsecond * 100)
}

type GraphObserver struct {
	concreteObserver Observer
}

func NewGraphObserver() *GraphObserver {
	graphObserver := &GraphObserver{}
	graphObserver.concreteObserver = graphObserver

	return graphObserver
}
